ding/example/ppo_with_complex_obs.py

```diff
@@ -0,0 +1,197 @@
+                    '2': spaces.Box(low=0, high=255, shape=image_size, dtype=np.uint8),
+                    '3': spaces.Box(low=0, high=np.array([np.inf, 3]), shape=(2, ), dtype=np.float32)
+                }
+            )
+        )
+
+    def reset(self):
+        # Generate a random initial state
+        return self.observation_space.sample()
+
+    def step(self, action):
+        # Compute the reward and done flag (which are not used in this example)
+        reward = np.random.uniform(low=0.0, high=1.0)
+
+        done = False
+        if np.random.uniform(low=0.0, high=1.0) > 0.7:
+            done = True
+
+        info = {}
+
+        # Return the next state, reward, and done flag
+        return self.observation_space.sample(), reward, done, info
+
+
+def ding_env_maker():
+    return DingEnvWrapper(
+        MyEnv(), cfg={'env_wrapper': [
+            lambda env: EvalEpisodeReturnEnv(env),
+        ]}
+    )
+
+
+class Encoder(nn.Module):
+
+    def __init__(self, feature_dim: int):
+        super(Encoder, self).__init__()
+
+        # Define the networks for each input type
+        self.fc_net_1_k1 = nn.Sequential(nn.Linear(1, 8), nn.ReLU())
+        self.fc_net_1_k2 = nn.Sequential(nn.Linear(1, 8), nn.ReLU())
+        self.fc_net_1 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+
+        self.class_token = nn.Parameter(torch.zeros(1, 1, feature_dim))
+        self.encoder_layer = nn.TransformerEncoderLayer(d_model=feature_dim, nhead=2, batch_first=True)
+        self.transformer_encoder = nn.TransformerEncoder(self.encoder_layer, num_layers=1)
+
+        self.conv_net = nn.Sequential(
+            nn.Conv2d(3, 16, kernel_size=3, padding=1), nn.ReLU(), nn.Conv2d(16, 32, kernel_size=3, padding=1),
+            nn.ReLU()
+        )
+        self.conv_fc_net = nn.Sequential(nn.Flatten(), nn.Linear(3200, 64), nn.ReLU())
+
+        self.fc_net_2 = nn.Sequential(nn.Linear(2, 16), nn.ReLU(), nn.Linear(16, 32), nn.ReLU(), nn.Flatten())
+
+    def forward(self, inputs: dict):
+        # Unpack the input tuple
+        dict_input = inputs['0']  # dict{key:(B)}
+        transformer_input = inputs['1']  # (B, seq_len, feature_dim)
+        conv_input = inputs['2']  # (B, H, W, 3)
+        fc_input = inputs['3']  # (B, X)
+
+        B = fc_input.shape[0]
+
+        # Pass each input through its corresponding network
+        dict_output = self.fc_net_1(
+            torch.cat(
+                [self.fc_net_1_k1(dict_input['k1'].unsqueeze(-1)),
+                 self.fc_net_1_k2(dict_input['k2'].unsqueeze(-1))],
+                dim=1
+            )
+        )
+
+        batch_class_token = self.class_token.expand(B, -1, -1)
+        transformer_output = self.transformer_encoder(torch.cat([batch_class_token, transformer_input], dim=1))
+        transformer_output = transformer_output[:, 0]
+
+        conv_output = self.conv_fc_net(self.conv_net(conv_input.permute(0, 3, 1, 2)).permute(0, 2, 3, 1))
```

Comment:
we don't need the second permute operation

Reply:
fix